keras_tuner/tuners/genetic.py

```diff
@@ -0,0 +1,542 @@
+            Defaults to "roulette_wheel".
+        seed: Optional integer, the random seed.
+        hyperparameters: Optional `HyperParameters` instance. Can be used to
+            override (or register in advance) hyperparameters in the search
+            space.
+        tune_new_entries: Boolean, whether hyperparameter entries that are
+            requested by the hypermodel but that were not specified in
+            `hyperparameters` should be added to the search space, or not. If
+            not, then the default value for these parameters will be used.
+            Defaults to True.
+        allow_new_entries: Boolean, whether the hypermodel is allowed
+            to request hyperparameters that were not specified in
+            `hyperparameters`. If not, then an error will be raised. Defaults
+            to True.
+    """
+
+    def __init__(
+        self,
+        objective=None,
+        generation_size=10,
+        population_size=10,
+        offspring_size=None,
+        mutation_factor=0.9,
+        crossover_factor=0.1,
+        threshold=None,
+        selection_type="tournament",
+        seed=None,
+        hyperparameters=None,
+        tune_new_entries=True,
+        allow_new_entries=True,
+    ):
+        self.generation_size = generation_size
+        self.population_size = population_size
+        self.offspring_size = offspring_size or population_size
+        self.max_trials = self._make_max_trials
+
+        super(GeneticOptimizationOracle, self).__init__(
+            objective=objective,
+            max_trials=self.max_trials,
+            hyperparameters=hyperparameters,
+            tune_new_entries=tune_new_entries,
+            allow_new_entries=allow_new_entries,
+            seed=seed,
+        )
+
+        self.mutation_factor = mutation_factor
+        self.crossover_factor = crossover_factor
+        self.threshold = threshold
+        self.selection_type = selection_type
+        self.seed = seed or random.randint(1, int(1e4))
+        self._seed_state = self.seed
+        self._random_state = np.random.RandomState(self.seed)
+
+        if self.mutation_factor + self.crossover_factor > 1.0:
+            raise ValueError(
+                "The sum of the 'mutation_factors' and 'crossover_factors' "
+                "must be less than 1.0."
+            )
+        if self.selection_type not in ["roulette_wheel", "tournament"]:
+            raise ValueError(
+                "The 'selection_type' must be either 'roulette_wheel' or "
+                "'tournament'."
+            )
+        self._tried_so_far = set()
+        self._max_collisions = 20
+        self.gep = self._make_gep()
+        self.ranges = self._make_ranges
+        self.population = {"hyperparameters": [], "scores": []}
+        self.new_population = {"hyperparameters": [], "scores": []}
+        self.values = {hp.name: hp.default for hp in self.get_space().space}
```

Comment:
We don't need this. It can be a local variable in `self.populate_space()` and an argument in `self._check_score(..., values)`.

Reply:
The problem is if you define in `self.populate_space()`, it actually starts to re-initialize the `hp.default` again and again for every trial. In general, this implementation is kind of nested because I was focusing on using the oracle `trial_id` for making use of the parallelisation so far.

Make it sense a little?